iis-wf/iis-wf-export-actionmanager/src/main/java/eu/dnetlib/iis/wf/export/actionmanager/module/CitationsActionBuilderModuleUtils.java

```diff
@@ -1,56 +1,132 @@
+        blobCitationEntry.setRawText(rawTextFn.apply(citationEntry));
+        blobCitationEntry.setIdentifiers(identifiersFn.apply(citationEntry));
+        return blobCitationEntry;
+    }
+
+    /**
+     * Build position value of blob citation entry from citation entry.
+     */
+    public static class PositionBuilder {
+        public static Integer build(CitationEntry citationEntry) {
+            return citationEntry.getPosition();
+        }
+    }
+
+    /**
+     * Build raw text value of blob citation entry from citation entry.
      */
-    public static BlobCitationEntry build(CitationEntry entry) {
-        BlobCitationEntry result = new BlobCitationEntry(
-                entry.getRawText() != null ? entry.getRawText().toString() : null);
-        result.setPosition(entry.getPosition());
-        if (entry.getDestinationDocumentId() != null) {
-            result.setIdentifiers(new ArrayList<TypedId>());
-            result.getIdentifiers()
-                    .add(new TypedId(entry.getDestinationDocumentId().toString(),
-                            ExtraInfoConstants.CITATION_TYPE_OPENAIRE,
-                            entry.getConfidenceLevel() != null
-                                    ? (entry.getConfidenceLevel() * InfoSpaceConstants.CONFIDENCE_TO_TRUST_LEVEL_FACTOR)
-                                    : 1f * InfoSpaceConstants.CONFIDENCE_TO_TRUST_LEVEL_FACTOR));
+    public static class RawTextBuilder {
+        public static String build(CitationEntry citationEntry) {
+            return Optional.ofNullable(citationEntry.getRawText()).map(CharSequence::toString).orElse(null);
         }
-        if (entry.getExternalDestinationDocumentIds() != null && !entry.getExternalDestinationDocumentIds().isEmpty()) {
-            if (result.getIdentifiers() == null) {
-                result.setIdentifiers(new ArrayList<TypedId>());
+    }
+
+    /**
+     * Build identifiers value of blob citation entry from citation entry.
+     */
+    public static class IdentifiersBuilder {
+        public static List<TypedId> build(CitationEntry citationEntry) {
+            return build(citationEntry,
+                    TypedIdFromDestinationDocumentIdBuilder::isValid,
+                    TypedIdFromDestinationDocumentIdBuilder::build,
+                    TypedIdsFromExternalDestinationDocumentIdsBuilder::isValid,
+                    TypedIdsFromExternalDestinationDocumentIdsBuilder::build);
+        }
+
+        public static List<TypedId> build(CitationEntry citationEntry,
+                                          Function<CitationEntry, Boolean> destinationDocumentIdValidator,
+                                          Function<CitationEntry, TypedId> typedIdFromDestinationDocumentIdFn,
+                                          Function<CitationEntry, Boolean> externalDestinationDocumentIdsValidator,
+                                          Function<CitationEntry, List<TypedId>> typedIdsFromExternalDestinationDocumentIdsFn) {
```

Comment:
Again, why so many `Function`s? Do we really need to express the logic this way? It is so difficult to track the code...

Reply:
Here, the reason was to allow for easy testing - I've wanted to separate the logic to make the test cases smaller and more isolated. As noted in other comment, I'm going to apply this pattern carefully in the future.